fabric-api-base/src/main/java/net/fabricmc/fabric/api/event/Event.java

```diff
@@ -43,6 +43,17 @@ public final T invoker() {
 		return invoker;
 	}
 
+	/**
+	 * Some events can be optimized away if nobody is listening. If an Event
+	 * implementation can quickly check for this condition, it may override this
+	 * method to return false in that case.
```

Comment:
I don't agree with this `hasListeners` method being necessary atm. What will occur is that if there is no invoker, the event will generate a sort of "dummy invoker" based on the lambda in the EventFactory.createArrayBacked(..., Function<..., ...>) thing.

If you are truly concerned about performance, you can specify an empty invoker manually via the other `createArrayBacked` method which uses your specified invoker to handle the event. Do note that the little overhead you get from this is not going to be very negligible unless it's something like events in rendering which specify the empty invoker for performance reasons.

Reply:
I'd personally say that it's out of scope for this PR.

Reply:
I'm not very concerned about performance, but since having no listeners on this event will be very common, might as well optimize for it.